test/js/bun/spawn/spawn.test.ts

```diff
@@ -837,3 +839,187 @@ it("error does not UAF", async () => {
   }
   expect(emsg).toInclude(" ");
 });
+
+describe("onDisconnect", () => {
+  it("onDisconnect callback is called when IPC disconnects", async () => {
+    const tmpDir = tmpdirSync();
+    const childScript = path.join(tmpDir, "ipc-disconnect-child.js");
+    writeFileSync(
+      childScript,
+      `
+        process.send("hello");
+        process.disconnect();
+        Promise.resolve().then(() => process.exit(0));
+      `,
+    );
+
+    let disconnectCalled = false;
+
+    const { promise, resolve } = Promise.withResolvers<void>();
+
+    await using proc = spawn({
+      cmd: [bunExe(), childScript],
+      ipc: (message, subprocess) => {
+        expect(message).toBe("hello");
+      },
+      onDisconnect: () => {
+        disconnectCalled = true;
+        expect(disconnectCalled).toBe(true);
+        resolve();
+      },
+      stdio: ["inherit", "inherit", "inherit"],
+      env: bunEnv,
+    });
+
+    await promise;
+  });
+
+  it("onDisconnect callback receives correct arguments", done => {
+    const tmpDir = tmpdirSync();
+    const childScript = path.join(tmpDir, "ipc-disconnect-child2.js");
+    writeFileSync(
+      childScript,
+      `
+        process.send("test");
+        process.disconnect();
+        Promise.resolve().then(() => process.exit(0));
+      `,
+    );
+
+    (async () => {
+      await using proc = spawn({
+        cmd: [bunExe(), childScript],
+        ipc: () => {},
+        onDisconnect: () => {
+          // onDisconnect should be called without arguments
+          done();
+        },
+        stdio: ["inherit", "inherit", "inherit"],
+        env: bunEnv,
+      });
+    })();
+  });
+
+  it("onDisconnect is not called when IPC is not used", async () => {
+    let disconnectCalled = false;
+
+    const proc = spawn({
+      cmd: [bunExe(), "-e", "console.log('hello')"],
+      onDisconnect: () => {
+        disconnectCalled = true;
+      },
+      stdout: "pipe",
+      stderr: "ignore",
+      stdin: "ignore",
+    });
+
+    await proc.exited;
+    expect(disconnectCalled).toBe(false);
+  });
+});
+
+describe("argv0", () => {
+  it("argv0 option changes process.argv0 but not executable", async () => {
+    const proc = spawn({
+      cmd: [bunExe(), "-e", "console.log(process.argv0); console.log(process.execPath)"],
+      argv0: "custom-argv0",
+      stdout: "pipe",
+      stderr: "ignore",
+      stdin: "ignore",
+      env: bunEnv,
+    });
+
+    const output = await proc.stdout.text();
+    const lines = output.trim().split(/\r?\n/);
+    expect(lines[0]).toBe("custom-argv0");
+    expect(path.normalize(lines[1])).toBe(path.normalize(bunExe()));
+    await proc.exited;
```

Comment:
I wonder if this would be cleaner with `jsonb`

```typescript
const testArgv0 = "custom-argv0";
const proc = spawn({
  cmd: [bunExe(), "-e", "console.log(JSON.stringify({ argv0: process.argv0, execPath: process.execPath }))"],
  argv0: testArgv0,
  stdout: "pipe",
  stderr: "ignore",
  stdin: "ignore",
  env: bunEnv,
});

const output = JSON.parse((await proc.stdout.text()).trim());
expect(output).toEqual({ argv0: testArgv0, execPath: path.normalize(bunExe()) });
```

I do suppose that's slightly more robust:

1. The nasty regex split is gone.
2. `custom-argv0` isn't repeated twice.